### DISCUSSION TITLE
Accessing a `Ref` when a `BuildContext` is available

### DISCUSSION THREAD

**Discussion by zepolyerf** [rrousselGit/riverpod#1937] [Q&A] [answered]
I'm in the process of migrating my app to use Riverpod. At the moment, my app is using the `provider` package to manage app state.

I have a lot of static methods inside multiple utility classes that take a `BuildContext` as input. I use the context to access the provider, which allows me to read and modify the state:

```dart
// This relies on the provider package
abstract class SomeRandomUtilityClass {

  static FutureOr<String?> redirect(BuildContext context) {
    bool isSignedIn = context.read<AuthProvider>().isSignedIn;
  
    if (!isSignedIn) {
      return '/signIn';
    } else {
      return null;
    }
  }
```

How do I achieve this with Riverpod? From what I understand, you need a `Ref` to be able to read/write from/to a provider, but I'm not even sure if the APIs allow it.

I tried replacing the `BuildContext` parameters of my methods with a `Ref` object, but I'm having trouble getting one. All I can access is a `WidgetRef`:

```dart
class App extends ConsumerWidget {
  /** ... **/

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return MaterialApp.router(
      /** ... **/
      routerConfig: GoRouter(
        /** ... **/
        routes: [
          GoRoute(
            /** ... **/
            // This doesn't work, because the method I'm calling expects a Ref,
            // and not a WidgetRef.
            redirect: (_, __) => SomeRandomUtilityClass.redirect(ref),
          ),
        ],
      ),
    );
  }
}
```

An easy fix would be to modify my static methods by changing the type of the parameter from `Ref` to `WidgetRef`, but it's been discussed here that [passing `WidgetRef`s is a bad idea](https://github.com/rrousselGit/riverpod/discussions/1188#discussioncomment-2172300).

I might be misunderstanding something. Please let me know if I am.

**Comment by rrousselGit** [comment, marked as answer]
You probably should change your static methods into class methods, and expose the class instance in a provider

```dart
class SomeUtility {
  SomeUtility(this.ref);
  final Ref ref;

  void redirect(...) {}
}

final someUtilityProvider = Provider(SomeUtility.new);

...


redirect: (_, __) => ref.read(someUtilityProvider).redirect()
```

**Reply by zepolyerf** [reply]
Thanks for the answer!
